[TOPIC-BLE-LLCP] Bluetooth: Controller: Conn Update Procedure

tosk-ot opened the pull request:
A minimum of the Connection Update Procedure and its unit test is added.
The procedure does not yet contain all handling of actual functionality.
Unit tests have been prepared for PDU data and added support for
selected or unselected BT_CTLR_CONN_PARAM_REQ as well as its feature
bit.

Signed-off-by: Tommie Skriver <tosk@demant.com>

tosk-ot committed:
tests: Bluetooth: LLCP: Add mock functions

Added a few mocks to be used by the unit tests.

Signed-off-by: Tommie Skriver <tosk@demant.com>

tosk-ot committed:
tests: Bluetooth: LLCP: Add various Kconfig settings for unit tests

Kconfigs referenced in the Connection Update Procedure.

Signed-off-by: Tommie Skriver <tosk@demant.com>

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_connections.c:
```diff
@@ -506,6 +516,16 @@ int ull_conn_llcp(struct ll_conn *conn, uint32_t ticks_at_expire, uint16_t lazy)
 {
 	LL_ASSERT(conn->lll.handle != ULL_HANDLE_NOT_CONNECTED);
 
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
```
I realized this is needed re. procedures that use instant (I had missed it re. phy update), so it should be passed in llcp and not llcp.conn_upd

tosk-ot replied:
Will move to llcp.

erbr-ot reviewed (commented):
I miss the update to ull_llcp_hci.c and respective unittest

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -545,20 +549,63 @@ uint8_t ull_cp_conn_update(struct ll_conn *conn, uint16_t interval_min, uint16_t
+#endif /* CONFIG_BT_CTLR_CONN_PARAM_REQ */
+	default:
+		/* CPR feature not supported */
+		return BT_HCI_ERR_CMD_DISALLOWED;
```
I agree

tosk-ot replied:
I agree, too! Will replace with an assert.

thoh-ot commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -1,557 +0,0 @@
-/*
```
Why is this file deleted in this commit?

tosk-ot replied:
This seems to be an unsuccessful attempt to move a file from one commit to another.
Will be fixed.

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:
```diff
@@ -137,10 +137,36 @@ struct proc_ctx {
+			uint16_t *pdu_win_offset0;
+			uint32_t ticks_ref;
+			uint32_t ticks_to_offset_next;
+		} conn_param;
```
Use shortform name `cpr`, so it matches the naming scheme of the rest of the structure

tosk-ot replied:
Done.

thoh-ot reviewed (changes requested):
There seems to be a general data structure problem, `llcp_struct::conn_upd` and `llcp_struct::conn_param` er just legacy place holders that should be transformed into the correct place.

tosk-ot commented on tests/bluetooth/controller/ctrl_conn_update/CMakeLists.txt:
```diff
@@ -10,6 +10,10 @@ FILE(GLOB SOURCES
   src/*.c
 )
 
+if(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+  add_compile_definitions(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+endif()
+
```
Yup. Will be removed.

tosk-ot commented on tests/bluetooth/controller/ctrl_conn_update/testcase.yaml:
```diff
@@ -2,4 +2,10 @@ common:
   bluetooth.controller.ctrl_conn_update.test:
-     type: unit
+    type: unit
+    timeout: 5 # override 60s default timeout
```
Yup. Will be removed.

tosk-ot commented on tests/bluetooth/controller/ctrl_conn_update/testcase.yaml:
```diff
@@ -2,4 +2,10 @@ common:
+
+  bluetooth.controller.ctrl_conn_update.no_param_req_test:
+    type: unit
+    timeout: 5 # override 60s default timeout
```
Yup. Will be removed.

tosk-ot commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -0,0 +1,576 @@
+/*
```
This seems to be an unsuccessful attempt to move a file from one commit to another.
Will be fixed.

tosk-ot committed:
tests: Bluetooth: LLCP: Add Kconfig override for Conn Update Proc.

Add Kconfig override file for Conn Update Procedure.

Signed-off-by: Tommie Skriver <tosk@demant.com>

tosk-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -765,3 +1075,76 @@ void ull_cp_priv_rp_conn_param_req_neg_reply(struct ll_conn *conn, struct proc_c
+
+	/* Get the terminate structure reserved in the connection context.
+	 * The terminate reason and connection handle should already be
+	 * populated before this mayfly function was scheduled.
+	 */
+	rx = (void *)&conn->terminate.node_rx;
+	LL_ASSERT(rx->hdr.link);
+	link = rx->hdr.link;
+	rx->hdr.link = NULL;
+
+	/* Enqueue the terminate towards ULL context */
+	ull_rx_put(link, rx);
+	ull_rx_sched();
+}
+
+static void ticker_start_conn_op_cb(uint32_t status, void *param)
+{
+	void *p;
+
+	LL_ASSERT(status == TICKER_STATUS_SUCCESS);
+
+	p = ull_update_unmark(param);
+	LL_ASSERT(p == param);
+}
+
+static void ticker_op_stop_cb(uint32_t status, void *param)
+{
+	uint32_t retval;
+	static memq_link_t link;
+	static struct mayfly mfy = {0, 0, &link, NULL, tx_lll_flush};
+
+	LL_ASSERT(status == TICKER_STATUS_SUCCESS);
+
+	mfy.param = param;
+
+	/* Flush pending tx PDUs in LLL (using a mayfly) */
+	retval = mayfly_enqueue(TICKER_USER_ID_ULL_LOW, TICKER_USER_ID_LLL, 0,
+				&mfy);
+	LL_ASSERT(!retval);
+}
```
Yup. Will use that instead.

tosk-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -394,12 +394,13 @@ struct llcp_struct {
 		uint16_t instant;
 		uint16_t *pdu_win_offset;
 		uint32_t ticks_anchor;
+		uint32_t ticks_at_expire;	/* TODO: Place correct! */
+		uint16_t lazy;			/* TODO(tosk): Place correct! */
```
Will be moved.

tosk-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -167,14 +181,252 @@ static void lp_cu_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
+			      TICKER_USER_ID_ULL_LOW, 1);
+	}
+#endif /* CONFIG_BT_CTLR_ULL_HIGH_PRIO == CONFIG_BT_CTLR_ULL_LOW_PRIO */
+}
```
Done.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -209,19 +234,31 @@ static void lp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
 	} else {
 		/* TODO(thoh): Hardcoded instant delta +6 */
-		ctx->data.cu.instant = lp_event_counter(conn) + 6;
+		ctx->data.cu.instant = lp_event_counter(conn) + 6U;
```
why not use the ull_conn_event_counter()? 
Also same comment as below re. define for delta

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -495,12 +537,23 @@ static void rp_cu_conn_param_req_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
-		ctx->state = RP_CU_STATE_WAIT_NTF;
-	} else {
-		rp_cu_ntf(conn, ctx);
+	/*
```
I'd prefer refactoring to not have complete-case code duplicated, but rather have a guard on the ntf to skip if not needed? Perhaps also change the prams_changed flag into a ntf flag - allowing the incorporation of the check for ctx->proc == PROC_CONN_UPDATE

tosk-ot replied:
As above, the check for ctx->proc == PROC_CONN_UPDATE does not apply and will be removed.
Agree with the duplicated code stuff - but this must come be later...

erbr-ot reviewed (commented):
WIP

tosk-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -169,12 +182,23 @@ static void lp_cu_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
+	 *
+	 * ParamReq: Notify Host.
+	 */
+	if ((ctx->proc == PROC_CONN_UPDATE) && (ctx->data.cu.params_changed == 0U)) {
```
The (ctx->proc == PROC_CONN_UPDATE) && will be removed,
as shall always apply here.

tosk-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c:
```diff
@@ -266,70 +268,21 @@ uint8_t ll_conn_update(uint16_t handle, uint8_t cmd, uint8_t status, uint16_t in
-			 *	conn->llcp_conn_param.cmd = 1U;
-			 */
+		if (status == 0U) {
+			ull_cp_conn_param_req_reply(conn);
```
I just realized that 'status' should be passed to ull_cp_conn_param_req_reply(), here:
Re.: 5.1.7.2: "If the Host rejects this request, then the device shall shall issue an LL_REJECT_EXT_IND PDU with the ErrorCode set to a value provided by the Host. The Host shall only use the error code Unacceptable Connection Parameters (0x3B) in order to reject the request."
So, it seems to be up to the Host to provide that error code...

tosk-ot committed:
Bluetooth: controller: Update HCI integr. of Connection Update Procedure

HCI integration of Connection Update Procedure is updated.
Unit test updated for compliance.

Signed-off-by: Tommie Skriver <tosk@demant.com>

tosk-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -391,34 +397,21 @@ struct llcp_struct {
 		uint32_t ticks_anchor;
 	} conn_upd;
 
+#if defined(CONFIG_BT_CTLR_LE_ENC)
```
Thus looks like must me messing up during resolving. Will be removed...

tosk-ot committed:
tests: Bluetooth: LLCP: Add Conn Update Procedure Unit Test

Added unit test test for Connection Update Procedure

Signed-off-by: Tommie Skriver <tosk@demant.com>

tosk-ot commented:
Rebased and updated

tosk-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -510,7 +563,7 @@ static void rp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
 	} else {
 		/* TODO(thoh): Hardcoded instant delta +6 */
-		ctx->data.cu.instant = rp_event_counter(conn) + 6;
+		ctx->data.cu.instant = rp_event_counter(conn) + 6U;
```
Define made. Maybe we should make it common at one point, as you mentioned.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -647,9 +766,24 @@ static void rp_cu_st_wait_rx_conn_update_ind(struct ll_conn *conn, struct proc_c
+			//TODO(tosk): skip/terminate if instant passed?
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+			/* conn param req procedure, if any, is complete */
+			conn->procedure_expire = 0U;
```
void ull_conn_prt_clear(struct ll_conn *conn);

Should probably be changed to an inline, however it still does the thing;-)

tosk-ot replied:
Done

erbr-ot reviewed (commented):
Review in progress - have to continue tomorrow

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_connections.c:
```diff
@@ -223,7 +225,247 @@ void ll_rx_enqueue(struct node_rx_pdu *rx)
+	return event_counter;
+}
+
+bool ull_conn_is_params_changed(struct ll_conn *conn, uint16_t interval,
```
Perhaps this should be in ull_llcp_conn_upd.c - so as to not 'load' ull_connections.c with cu specific code?
Perhaps we should have a discussion on this on Thursday?

tosk-ot replied:
Agree! Moved (and renamed)

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_connections.c:
```diff
@@ -223,7 +225,247 @@ void ll_rx_enqueue(struct node_rx_pdu *rx)
+	instant_latency = (event_counter - instant) &
+			  0xFFFF;
+
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
```
I would put this outside of this function - and use the API as mentioned in other comment

erbr-ot replied:
or place it next to the apto/sto handling below

tosk-ot replied:
I'm not sure I know what you mean?

erbr-ot replied:
No, I see - looks like the comment was put on the wrong line

erbr-ot replied:
I'm thinking of the handling of conn->procedure_expire, I'd move that code to where the other timeout related functionality is in this function (re. apto and sto)

erbr-ot replied:
But its a nit pick!

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -219,6 +222,7 @@ struct proc_ctx *ull_cp_priv_create_local_procedure(enum llcp_proc proc)
 	default:
 		/* Unknown procedure */
 		LL_ASSERT(0);
+		break;
```
Is this break due to personal preference or coding guide lines?
Asking as we are not doing this elsewhere, so perhaps we should?

tosk-ot replied:
Misra.
https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_16_03.c

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -183,8 +216,19 @@ static void lp_cu_send_conn_param_req(struct ll_conn *conn, struct proc_ctx *ctx
 		rr_set_incompat(conn, INCOMPAT_RESOLVABLE);
 
+		ctx->data.cu.reference_conn_event_count = event_counter;
+		ctx->data.cu.preferred_periodicity = 0U;
```
I'd personally prefer putting this initialization in a inline function or perhaps move it to init of the procedure.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -209,19 +254,30 @@ static void lp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
+
 		lp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND);
-		ctx->rx_opcode = 0xFF; /* TODO(thoh): Hmm */
+		ctx->rx_opcode = PDU_DATA_LLCTRL_TYPE_UNUSED; /* TODO(thoh): Hmm */
```
remove the 'hmm' comment?

tosk-ot replied:
Yes.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -510,19 +595,19 @@ static void rp_cu_send_conn_update_ind(struct ll_conn *conn, struct proc_ctx *ct
+		ctx->data.cu.instant = rp_event_counter(conn) + CONN_UPDATE_INSTANT_DELTA;
 		rp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND);
-		ctx->rx_opcode = 0xFF; /* TODO(thoh): Hmm */
+		ctx->rx_opcode = PDU_DATA_LLCTRL_TYPE_UNUSED; /* TODO(thoh): Hmm */
```
'hmm'

tosk-ot replied:
Yes.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:
```diff
@@ -566,7 +672,19 @@ static void rp_cu_st_wait_rx_conn_param_req(struct ll_conn *conn, struct proc_ct
-		rp_cu_send_conn_param_req_ntf(conn, ctx, evt, param);
+		pdu_decode_conn_param_req(ctx, param);
+
+		bool parameters_is_changed = ull_conn_is_params_changed(conn,
```
Nit-pick (capital N) spelling: parameters_ARE_changed or better jet just loose the 'is' (also in function!)

tosk-ot replied:
Sure! Function and parameters renamed!

tosk-ot commented:
Rebased, updated and fixed compliance errors.

tosk-ot committed:
Bluetooth: Controller: Conn Update Procedure

Add initial functionality for the Connection update Procedure.

The procedure does not yet contain all handling of actual functionality.

Signed-off-by: Tommie Skriver <tosk@demant.com>

erbr-ot reviewed (approved):
LGTM
